l5x/type.go

```diff
@@ -0,0 +1,380 @@
+	}
+
+	if mb.Dimension > 1 {
+		nt.Type = arrayType{
+			elementInfo: nt.Type,
+			count:       mb.Dimension,
+		}
+	}
+
+	return nt, nil
+}
+
+func (dt DataType) AsType(knownTypes TypeList) (Type, error) {
+	if dt.Class != ClassUser {
+		return nil, fmt.Errorf("Unknown class type")
+	}
+
+	switch dt.Family {
+	case DataTypeFamilyString:
+		return parseString(dt.Name, dt.Members)
+
+	case DataTypeFamilyNone:
+		return parseStruct(dt.Name, dt.Members, knownTypes)
+
+	default:
+		return nil, fmt.Errorf("Unknown data family type")
+	}
+}
+
+func parseStruct(name string, membs []Member, knownTypes TypeList) (Type, error) {
+	sti := structType{
+		name:    name,
+		members: []NamedType{},
+	}
+	for _, memb := range membs {
+		if memb.DataType == "BIT" {
+			// Not yet implemented, so ignore it
+			continue
+		}
+		nm, err := memb.AsNamedType(knownTypes)
+		if err != nil {
+			if errors.Is(err, ErrUnknownType) {
+				err = errUnknownTypeSpecific{name, memb.DataType}
+			}
+			return nil, err
+		}
+		sti.members = append(sti.members, nm)
+	}
+	if len(sti.members) == 0 {
+		return nil, fmt.Errorf("DataType '%s' produced no members (%d)", name, len(membs))
+	}
+
+	return sti, nil
+}
+
+func parseString(name string, memb []Member) (Type, error) {
+	if len(memb) != 2 {
+		return nil, fmt.Errorf("StringFamily '%s' had %d members instead of 2", name, len(memb))
+	}
+
+	if memb[0].Name != "LEN" {
+		return nil, fmt.Errorf("StringFamily '%s' LEN member is missing: %s", name, memb[0].Name)
+	}
+	if memb[0].DataType != "DINT" {
+		return nil, fmt.Errorf("StringFamily '%s' LEN.DataType is incorrect: %v", name, memb[0].DataType)
+	}
+	if memb[0].Dimension != 0 {
+		return nil, fmt.Errorf("StringFamily '%s' LEN.Dimension is incorrect: %d", name, memb[0].Dimension)
+	}
+	if memb[0].Radix != RadixDecimal {
+		return nil, fmt.Errorf("StringFamily '%s' LEN.Radix is incorrect: %v", name, memb[0].Radix)
+	}
+
+	if memb[1].Name != "DATA" {
+		return nil, fmt.Errorf("StringFamily '%s' DATA member is missing: %v", name, memb[1].Name)
+	}
+	if memb[1].DataType != "SINT" {
+		return nil, fmt.Errorf("StringFamily '%s' DATA.DataType is incorrect: %v", name, memb[1].DataType)
+	}
+	if memb[1].Dimension <= 0 {
+		return nil, fmt.Errorf("StringFamily '%s' DATA.Dimension is invalid: %d", name, memb[1].Dimension)
+	}
+	if memb[1].Radix != RadixASCII {
+		return nil, fmt.Errorf("StringFamily '%s' DATA.Radix is incorrect: %v", name, memb[1].Radix)
+	}
+
+	return stringType{
+		name: name,
+		atype: arrayType{
+			elementInfo: basicType{"SINT", "int8"},
+			count:       memb[1].Dimension,
+		},
+	}, nil
+}
+
+type basicType struct {
+	plcName, goName string
+}
+
+func (bt basicType) PlcName() string      { return bt.plcName }
+func (bt basicType) GoName() string       { return bt.goName }
+func (bt basicType) GoTypeString() string { return "" }
+
+type arrayType struct {
+	elementInfo Type
+	count       int
+}
+
+func (ati arrayType) PlcName() string { return "" }
```

Comment:
Is returning the empty string the right thing to do here?

Reply:
I think that was intentional. I believe code elsewhere checks and if this is an empty string, then it means the type is a base type and therefore already exists in go, so it doesn't need to be defined. Or something to that effect.